### PR TITLE
Get correct ledger state in block production

### LINKED ISSUE #1437: When creating a fork, mempool is evaluated wrt to wrong ledger state
If two leaders are elected for the same slot, we _could_ get the block from the other node before we manage to produce ours. When this happens, we should create a fork, connecting our new block to the _previous_ block, rather than appending it to the block we received. 

We currently do this, but we evaluate the mempool with respect to the wrong ledger state, so that we might produce an invalid block. Instead, we should evaluate it wrt to the previous ledger state.

This has a few consequences:

* When we applied the incoming block, we will likely have removed some transactions from our mempool, and so we will not include them in our own block.
  
  This will be mitigated to some degree by the thread that monitors the chain and reintroduces transactions that get removed by a switch-to-fork (#1269).
  
  However, even with that in place, we might still want to consider doing something special here. As it stands, that thread would notice some transactions got lost _after_ we produce our block, so that they might be included in the _next_. Ideally however we would include them in _this_ block.

  Note that from the perspective of other nodes, the situation is slightly different. They will either adopt the block produced by the other node, in which case anything we decide to do is irrelevant, or they will produce the block that we produce, in which case any transactions that may be missing will probably still exist in their mempool (assuming that all transactions are sent to all core nodes).
  
* A related problem is that the mempool can contain dependent transactions. After applying the incoming block, the mempool might now contain transactions that are valid with respect to the ledger state after applying the incoming block, but will be _invalid_ when evaluated against the previous ledger state (since the transactions they depend on have not been applied in that state). This means that when we evaluate the mempool with respect to that prior ledger state, those dependent transactions from the mempool will be removed as invalid, and not be included in the block we produce.

   **This is not addressed by the thread mentioned in the previous point**: the transactions that are deleted from the mempool are not included in the incoming block, _nor_ in our block, and so we would not reintroduce them into the mempool simply if we monitor the chain.
   
   From the perspective of another node, the situation is again slightly different. As above, if they don't end up adopting our block, what we decide here doesn't really matter. If they do, they simply wouldn't remove _either_ the transactions that had been included in the block produced by the other node, _nor_ the ones that we removed because they depended on them, and so they would be able to include all of them in the next block. This does however depend on other codes having the same transactions in their mempool.

* In order to be to correctly evaluate the mempool in this case, we must be able to evaluate it against the previous ledger state. This means that it's relatively important that we can efficiently get that previous ledger from the chain DB (#1026).

### PR DESCRIPTION
**Pull request by edsko**
This also changes tracing during block production; see `TraceForgeEvent` for details.

Closes #1437.

### CODE REVIEW

File: ouroboros-consensus/src/Ouroboros/Consensus/NodeKernel.hs [hunk 1]
```diff
@@ -310,100 +311,126 @@ forkBlockProduction
...
+        trace $ TraceForgeEvent currentSlot newBlock
+
+        -- Add the block to the chain DB
+        lift $ ChainDB.addBlock chainDB newBlock
+
+        -- Check whether we adopted our block
+        --
+        -- addBlock is synchronous, so when it returns the block we produced
+        -- will have been considered and possibly (probably) adopted
+        curTip <- lift $ atomically $ ChainDB.getTipPoint chainDB
```
Comment (edsko) [inline ouroboros-consensus/src/Ouroboros/Consensus/NodeKernel.hs]: We could have adopted more blocks on top of this

Comment (edsko) [reply ouroboros-consensus/src/Ouroboros/Consensus/NodeKernel.hs]: Have addBlock return the new tip?

File: ouroboros-consensus/src/Ouroboros/Consensus/NodeKernel.hs [hunk 2]
```diff
@@ -310,100 +311,126 @@ forkBlockProduction
...
+        -- may not be adopted, but it won't be invalid.
+        mempoolSnapshot <- lift $ atomically $ getSnapshotFor
+                                                 mempool
+                                                 (TxsForBlockInSlot currentSlot)
+                                                 (ledgerState extLedger)
+        let blockEncOverhead = nodeBlockEncodingOverhead ledger
+            maxBlockBodySize = case maxBlockSizeOverride of
+              NoOverride            -> nodeMaxBlockSize ledger - blockEncOverhead
+              MaxBlockSize mbs      -> mbs - blockEncOverhead
+              MaxBlockBodySize mbbs -> mbbs
```
Comment (edsko) [inline ouroboros-consensus/src/Ouroboros/Consensus/NodeKernel.hs]: Current configuration (max block size) could become out of date with max block size on the ledger; should take a `min` here.

File: ouroboros-consensus/src/Ouroboros/Consensus/NodeKernel.hs [hunk 3]
```diff
@@ -310,100 +311,126 @@ forkBlockProduction
...
+        -- NOTE: It is possible that due to adoption of new blocks the
+        -- /current/ ledger will have changed. This doesn't matter: we will
+        -- produce a block that fits onto the ledger we got above; if the
+        -- ledger in the meantime changes, the block we produce here may or
+        -- may not be adopted, but it won't be invalid.
+        mempoolSnapshot <- lift $ atomically $ getSnapshotFor
+                                                 mempool
+                                                 (TxsForBlockInSlot currentSlot)
+                                                 (ledgerState extLedger)
+        let blockEncOverhead = nodeBlockEncodingOverhead ledger
```
Comment (edsko) [inline ouroboros-consensus/src/Ouroboros/Consensus/NodeKernel.hs]: Trace size of mempool snapshot here

File: ouroboros-consensus/src/Ouroboros/Consensus/NodeKernel.hs (hunk 2 above)
Comment (edsko) [reply ouroboros-consensus/src/Ouroboros/Consensus/NodeKernel.hs]: (Make it part of TraceForgeEvent)

File: ouroboros-consensus/src/Ouroboros/Consensus/NodeKernel.hs (hunk 1 above)
Comment (edsko) [reply ouroboros-consensus/src/Ouroboros/Consensus/NodeKernel.hs]: Created a separate ticket for this at https://github.com/input-output-hk/ouroboros-network/issues/1463 .

**Comment by nfrisby** [comment]
Mini-review: The changes to these tracers seems compatible with `test-consensus`. 👍

File: ouroboros-consensus/src/Ouroboros/Consensus/NodeKernel.hs [hunk 4]
```diff
@@ -310,100 +311,152 @@ forkBlockProduction
...
+              exitEarly
+
+        -- Get ledger state corresponding to prevPoint
+        --
+        -- Under normal circumstances this should not fail. If we are unable to
+        -- get this ledger state, it means we switch to another chain which is
+        -- so distant or so much longer that the point we just looked up isn't
+        -- within @k@ blocks from the tip anymore. It's a theoretical possibility,
+        -- but very unlikely. When it happens, we simply give up on the chance
+        -- to produce a block.
```
Comment (dcoutts) [inline ouroboros-consensus/src/Ouroboros/Consensus/NodeKernel.hs]: switch/switched

Isn't there the rather straightforward possibility of switching to a different short fork that does not include `prevPoint`? For example roll back 1, roll forward 2, between calling `ChainDB.getCurrentChain` and `ChainDB.getPastLedger`

Comment (edsko) [reply ouroboros-consensus/src/Ouroboros/Consensus/NodeKernel.hs]: Ah, yes, you are right. I will update the comment.

**Pull request merged**